### PR TITLE
Saml changes

### PR DESCRIPTION
**Pull request by psantos1113**
Implement SAML changes and allow the feature to be turned on or off multiple times.
Fix Alert component circular references.

### CODE REVIEW

File: src/main/java/com/synopsys/integration/alert/Application.java [hunk 1]
```diff
@@ -84,72 +68,11 @@ void init() {
...
-        return new AuthenticationSupport();
-    }
-
-    @Bean
-    public HttpSessionCsrfTokenRepository csrfTokenRepository() {
-        return new HttpSessionCsrfTokenRepository();
-    }
-
     @Bean
     public DaoAuthenticationProvider alertDatabaseAuthProvider(final PasswordEncoder defaultPasswordEncoder) {
```
Comment (jamesrichard91) [inline src/main/java/com/synopsys/integration/alert/Application.java]: Do we still want this here? or inApplicationConfiguration?

File: src/main/java/com/synopsys/integration/alert/web/security/AuthenticationHandler.java [hunk 2]
```diff
@@ -22,41 +22,380 @@
...
+        };
+        return matchers;
+    }
+
+    private RequestMatcher[] createAllowedPathMatchers() {
+        final RequestMatcher[] matchers = {
+            new SamlAntMatcher(samlContext(), httpPathManager.getSamlAllowedPaths(), httpPathManager.getAllowedPaths())
+        };
+        return matchers;
+    }
```
Comment (bamandel) [inline src/main/java/com/synopsys/integration/alert/web/security/AuthenticationHandler.java]: This class is getting pretty large. Would we be able to separate out the Beans from here into a different configuration class?

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLMetadataGenerator.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    public String getEntityBaseURL() {
+        try {
+            final ConfigurationModel currentConfiguration = samlContext.getCurrentConfiguration();
+            return samlContext.getFieldValueOrEmpty(currentConfiguration, SettingsDescriptor.KEY_SAML_ENTITY_BASE_URL);
+        } catch (final AlertDatabaseConstraintException | AlertLDAPConfigurationException e) {
+            logger.error("Could not get the SAML entity base URL.", e);
+        }
+        return "";
+    }
```
Comment (bamandel) [inline src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLMetadataGenerator.java]: These two methods can probably use another helper method to cut out redundant code.

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertWebSSOProfileOptions.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public Boolean getForceAuthN() {
+        try {
+            final ConfigurationModel currentConfiguration = samlContext.getCurrentConfiguration();
+            return samlContext.getFieldValueBoolean(currentConfiguration, SettingsDescriptor.KEY_SAML_FORCE_AUTH);
+        } catch (final AlertDatabaseConstraintException | AlertLDAPConfigurationException e) {
+            logger.error("Could not get the SAML force AuthN.", e);
+        }
+        return false;
+    }
```
Comment (bamandel) [inline src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertWebSSOProfileOptions.java]: If we end up doing this sort of format a lot, then perhaps we want a helper class to cut back on all this duplicate error handling.

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SamlAntMatcher.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Override
+    public boolean matches(final HttpServletRequest request) {
+        final Collection<RequestMatcher> requestMatchers;
+        if (context.isSAMLEnabled()) {
+            requestMatchers = enabledMatchers;
+        } else {
+            requestMatchers = disabledMatchers;
+        }
```
Comment (bamandel) [inline src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SamlAntMatcher.java]: This could be simplified by defaultin requestMatchers to disabled then checking if it's enabled and changing the value if that's the case.

File: alert-common/src/main/java/com/synopsys/integration/alert/common/AlertProperties.java [hunk 6]
```diff
@@ -26,9 +26,7 @@
 
 import org.apache.commons.lang3.StringUtils;
 import org.springframework.beans.factory.annotation.Value;
-import org.springframework.stereotype.Component;
 
-@Component
```
Comment (gkillough) [inline alert-common/src/main/java/com/synopsys/integration/alert/common/AlertProperties.java]: Why was this deleted?

File: src/main/java/com/synopsys/integration/alert/component/settings/SettingsUIConfig.java [hunk 7]
```diff
@@ -84,6 +89,11 @@
     private static final String SETTINGS_LDAP_GROUP_SEARCH_BASE_DESCRIPTION = "The part of the LDAP directory in which group searches should be done.";
     private static final String SETTINGS_LDAP_GROUP_SEARCH_FILTER_DESCRIPTION = "The filter used to search for group membership.";
     private static final String SETTINGS_LDAP_GROUP_ROLE_ATTRIBUTE_DESCRIPTION = "The ID of the attribute which contains the role name for a group.";
+    private static final String SETTINGS_SAML_ENABLED_DESCRIPTION = "If true, Alert with attempt to authenticate using the SAML configuration.";
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/component/settings/SettingsUIConfig.java]: I think there is a typo here. "with" should be "will"

File: src/main/java/com/synopsys/integration/alert/web/controller/HomeController.java [hunk 8]
```diff
@@ -53,7 +55,14 @@ public String index() {
     public ResponseEntity<String> checkAuthentication(final HttpServletRequest request) {
         final HttpServletRequest httpRequest = request;
         final CsrfToken csrfToken = csrfTokenRespository.loadToken(request);
-        if (csrfToken == null) {
+        final Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        boolean authorized = false;
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/web/controller/HomeController.java]:
```java
final boolean authorized = authentication.isAuthenticated() && csrfToken != null;
```

File: src/main/java/com/synopsys/integration/alert/web/security/AuthenticationHandler.java (hunk 2 above)
Comment (gkillough) [reply src/main/java/com/synopsys/integration/alert/web/security/AuthenticationHandler.java]: AuthenticationConfiguration

File: src/main/java/com/synopsys/integration/alert/web/security/EnabledSSLValidator.java [hunk 9]
```diff
@@ -22,27 +22,15 @@
...
 import org.springframework.context.annotation.Profile;
-import org.springframework.security.config.annotation.web.builders.HttpSecurity;
-import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
-import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.stereotype.Component;
 
-@EnableWebSecurity
-@Configuration
+@Component
 @Profile("ssl")
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/web/security/EnabledSSLValidator.java]: Is there a constant we could use here or create?

File: src/main/java/com/synopsys/integration/alert/web/security/AuthenticationHandler.java (hunk 2 above)
Comment (psantos1113) [reply src/main/java/com/synopsys/integration/alert/web/security/AuthenticationHandler.java]: Unfortunately you can't because to create one of the SAML beans you need to call a method that only exists in this class.  I tried to separate it out once already to do exactly what you suggested.  Line 221 the authenticationManager() method is the issue.

File: src/main/java/com/synopsys/integration/alert/web/security/NoSSLValidator.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.synopsys.integration.alert.web.security;
+
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Component;
+
+@Component
+@Profile("no_ssl")
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/web/security/NoSSLValidator.java]: Constant?

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLEntryPoint.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        this.samlContext = samlContext;
+    }
+
+    @Override
+    protected boolean processFilter(final HttpServletRequest request) {
+        return samlContext.isSAMLEnabled() && super.processFilter(request);
+    }
+
+    @Override
+    public void commence(final HttpServletRequest request, final HttpServletResponse response, final AuthenticationException e) throws IOException, ServletException {
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLEntryPoint.java]: Will this only be called if we already know that SAML is enabled?

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLMetadataGenerator.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        try {
+            final ConfigurationModel currentConfiguration = samlContext.getCurrentConfiguration();
+            return samlContext.getFieldValueOrEmpty(currentConfiguration, SettingsDescriptor.KEY_SAML_ENTITY_ID);
+        } catch (final AlertDatabaseConstraintException | AlertLDAPConfigurationException e) {
+            logger.error("Could not get the SAML entity ID.", e);
+        }
+        return "";
+    }
+
+    public String getEntityBaseURL() {
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLMetadataGenerator.java]: Should this be optional?

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLMetadataGenerator.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+public class AlertSAMLMetadataGenerator extends MetadataGenerator {
+    private static final Logger logger = LoggerFactory.getLogger(AlertSAMLMetadataGenerator.class);
+    private final SAMLContext samlContext;
+
+    public AlertSAMLMetadataGenerator(final SAMLContext samlContext) {
+        this.samlContext = samlContext;
+    }
+
+    public String getEntityId() {
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLMetadataGenerator.java]: Should this be Optional?

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SAMLContext.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+                   .orElseThrow(() -> new AlertLDAPConfigurationException("Settings configuration missing"));
+    }
+
+    public boolean isSAMLEnabled() {
+        final boolean enabled = false;
+        try {
+            return isSAMLEnabled(getCurrentConfiguration());
+        } catch (final AlertDatabaseConstraintException | AlertLDAPConfigurationException ex) {
+            logger.warn(ex.getMessage());
+            logger.debug("cause: ", ex);
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SAMLContext.java]: Considering there is a colon here, I think you might have forgotten `"{}"`. I'm not sure how the logger handles this case if not.

File: src/main/java/com/synopsys/integration/alert/Application.java (hunk 1 above)
Comment (psantos1113) [reply src/main/java/com/synopsys/integration/alert/Application.java]: StartupManager cannot be autowired correctly if I try to separate it out.  The StartupManager class has too many dependencies that we need to look into refactoring.  This way the configuration is either in ApplicationConfiguration or AuthenticationHandler

**Comment by bamandel** [comment]
Do we want to merge these changes into master or into the 4.1.0 branch?

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SamlAntMatcher.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+    public SamlAntMatcher(final SAMLContext context, final String[] samlEnabledPatterns, final String[] samlDisabledPattern) {
+        this.context = context;
+        this.enabledPatterns = Set.of(samlEnabledPatterns);
+        this.disabledPatterns = Set.of(samlDisabledPattern);
+        createAntMatchers();
+    }
+
+    private void createAntMatchers() {
+        enabledMatchers = enabledPatterns.stream().map(pattern -> new AntPathRequestMatcher(pattern)).collect(Collectors.toList());
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SamlAntMatcher.java]: You might be able to use a method reference here: `AntPathRequestMatcher::new`.

File: src/main/js/LogoutPage.js [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    render() {
+        return (
+            <div className="wrapper">
+                <div className="loginContainer">
+                    <div className="loginBox">
+                        <Header />
+                        <div className="form-horizontal loginForm">
+                            <div className="row">
+                                <div className="col-sm-12 text-center">
+                                    <div className="d-inline-flex flex-column p-2">You've successfully logged out of Alert!</div>
```
Comment (gkillough) [inline src/main/js/LogoutPage.js]: Do we need to escape the ' character here?

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SAMLContext.java (hunk 14 above)
Comment (psantos1113) [reply src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SAMLContext.java]: The logger has an API method for Throwable as a second parameter.  So you don't need {} when logging an exception

File: src/main/js/LogoutPage.js (hunk 16 above)
Comment (psantos1113) [reply src/main/js/LogoutPage.js]: I will double check it looked fine in the UI because it is in a JSX section the raw html will be interpreted as html.  I will check though

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SamlAntMatcher.java (hunk 15 above)
Comment (psantos1113) [reply src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/SamlAntMatcher.java]: good catch I will try that.

File: src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLMetadataGenerator.java (hunk 12 above)
Comment (psantos1113) [reply src/main/java/com/synopsys/integration/alert/web/security/authentication/saml/AlertSAMLMetadataGenerator.java]: I will look into making those optional
